src/layout/umap.c

```diff
@@ -0,0 +1,1038 @@
+
+#ifdef UMAP_DEBUG
+    igraph_real_t cross_entropy, cross_entropy_old;
+#endif
+
+    /* Explicit avoidance of neighbor repulsion, only useful in small graphs
+     * which are never very sparse */
+    igraph_bool_t avoid_neighbor_repulsion = 0;
+    if (igraph_vcount(graph) < 100) {
+        avoid_neighbor_repulsion = 1;
+    }
+
+    /* Measure the (variable part of the) cross-entropy terms for debugging:
+     * 1. - sum_edge_e mu(e) * log(nu(e))
+     * 2. + sum_edge_e (1 - mu(e)) * log(1 - nu(e))
+     * The latter is approximated by negative sampling as:
+     * 2b. + sum_random_ij 1 * log(1 - nu_ij)
+     * whereby the mu = 0 because we assume there's no edge between i and j, and nu_ij
+     * is basically their distance in embedding space, lensed through the probability
+     * function Phi.
+     * */
+#ifdef UMAP_DEBUG
+    igraph_i_umap_compute_cross_entropy(graph, umap_weights, layout, a, b, &cross_entropy);
+#endif
+
+    for (igraph_integer_t e = 0; e < epochs; e++) {
+        /* Apply (stochastic) forces */
+        igraph_i_umap_apply_forces(graph, umap_weights, layout, a, b, sampling_prob, learning_rate,
+                avoid_neighbor_repulsion);
+
+#ifdef UMAP_DEBUG
+        /* Recompute CE and check how it's going*/
+        cross_entropy_old = cross_entropy;
+        igraph_i_umap_compute_cross_entropy(graph, umap_weights, layout, a, b, &cross_entropy);
+
+        printf("Cross-entropy before shift: %g, after shift: %g\n", cross_entropy_old, cross_entropy);
+#endif
+
+         /* Adjust learning rate */
+        learning_rate = 1.0 - (float)(e + 1) / epochs;
+    }
+
+    return IGRAPH_SUCCESS;
+}
+
+/* Center layout around (0,0) at the end, just for convenience */
+static igraph_error_t igraph_i_umap_center_layout(igraph_matrix_t *layout) {
+    igraph_integer_t no_of_nodes = igraph_matrix_nrow(layout);
+    igraph_real_t xm = 0, ym = 0;
+
+    /* Compute center */
+    xm = 0;
+    ym = 0;
+    for (igraph_integer_t i = 0; i < no_of_nodes; i++) {
+        xm += MATRIX(*layout, i, 0);
+        ym += MATRIX(*layout, i, 1);
+    }
+    xm /= no_of_nodes;
+    ym /= no_of_nodes;
+
+    /* Shift vertices */
+    for (igraph_integer_t i = 0; i < no_of_nodes; i++) {
+        MATRIX(*layout, i, 0) -= xm;
+        MATRIX(*layout, i, 1) -= ym;
+    }
+
+    return IGRAPH_SUCCESS;
+}
+
+
+/* Check the distances argument, which should be NULL or a vector of nonnegative numbers */
+static igraph_error_t igraph_i_umap_check_distances(const igraph_vector_t *distances, igraph_integer_t no_of_edges) {
+
+    if (distances == NULL) {
+        return IGRAPH_SUCCESS;
+    }
+
+    if (igraph_vector_size(distances) != no_of_edges) {
+        IGRAPH_ERROR("Distances must be the same number as the edges in the graph.", IGRAPH_EINVAL);
+    }
+
+    for (igraph_integer_t eid = 0; eid != no_of_edges; eid++) {
+        if (VECTOR(*distances)[eid] < 0) {
+            IGRAPH_ERROR("Distances cannot be negative.", IGRAPH_EINVAL);
+        }
+    }
```

Comment:
We normally use `vector_min` for such checks, but pay attention that it cannot be called on empty vectors. At the same time, check for NaN.  Here's e.g. how it's done by `closeness()`:

```c
    if (no_of_edges > 0) {
        igraph_real_t minweight = igraph_vector_min(weights);
        if (minweight <= 0) {
            IGRAPH_ERROR("Weight vector must be positive.", IGRAPH_EINVAL);
        } else if (igraph_is_nan(minweight)) {
            IGRAPH_ERROR("Weight vector must not contain NaN values.", IGRAPH_EINVAL);
        }
    }
```

Reply:
no need to go through the whole vector if the first or second value is negative (a likely scenario, if the user just got the wrong vector in). I added the NaN check